### PR TITLE
[flink]Supported ogg format and optimize format scalability

### PR DESCRIPTION
**Pull request by MonsterChenzhuo**
### Purpose
subtask issue: #1826

This PR accomplishes the following:
It completes the decoupling work of the format. Through the abstraction of DataFormat, integrating a new format has been simplified.

For instance, if you wish to develop a new format, you only need to follow two steps, without altering or modifying the existing code logic:

Implement the parsing logic for the new format. You can refer to my implementation at: org/apache/paimon/flink/action/cdc/kafka/formats/ogg/OggRecordParser.java.
Add this new format to DataFormat.

### Tests
IT：KafkaOggSyncDatabaseActionITCase
图片

IT：KafkaOggSyncTableActionITCase
图片

### CODE REVIEW

**Review by JingsongLi — COMMENTED** [review]
Very thanks to @MonsterChenzhuo .

This is what we want to do! We must do some abstraction for log sync.
1. Abstraction for format: canal debezium and more...
2. Abstraction for log system: kafka pulsar..

This is the first step.

**Review by JingsongLi — COMMENTED** [review]
Can you try to implement a parser for debezium?
We should take a look if this refactor is OK for other formats.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/formats/DataFormat.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+            return DataFormat.valueOf(formatStr.replace("-", "_").toUpperCase());
+        } catch (Exception e) {
+            throw new UnsupportedOperationException(
+                    "This format: " + formatStr + " is not support.");
+        }
+    }
+
+    /** Data format formats. */
+    @FunctionalInterface
+    public interface DataFormatParser {
```
Comment (yuzelin) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/formats/DataFormat.java:60]: I think this interface can be named `RecordParserFactory`.

**Comment by MonsterChenzhuo** [comment]
> Can you try to implement a parser for debezium? We should take a look if this refactor is OK for other formats.

Sure, I can do that.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/formats/DataFormat.java (hunk 1 above)
Comment (MonsterChenzhuo) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/formats/DataFormat.java]: done.

**Comment by MonsterChenzhuo** [comment]
> Very thanks to @MonsterChenzhuo .
> 
> This is what we want to do! We must do some abstraction for log sync.
> 
>     1. Abstraction for format: canal debezium and more...
> 
>     2. Abstraction for log system: kafka pulsar..
> 
> 
> This is the first step.
@JingsongLi @yuzelin  PTAL,thanks.

**Comment by Alibaba-HZY** [comment]
very well job,ogg-json does not contain field type information, all types are string, whether this does not meet user expectation? but there doesn't seem to be a better solution

**Comment by MonsterChenzhuo** [comment]
> very well job,ogg-json does not contain field type information, all types are string, whether this does not meet user expectation? but there doesn't seem to be a better solution

In my personal experience, string types work for most situations. However, I have a follow up plan for your concern.
I plan to implement it later:
1. In the whole library synchronization scenario, allow users to create tables in Paimon first. Synchronization will be based on the table created by the user for type mapping.
2. Table level synchronization. This will allow users to pass data type parameters. This approach is similar to FlinkSQL.

**Comment by MonsterChenzhuo** [comment]
> Can you try to implement a parser for debezium? We should take a look if this refactor is OK for other formats.



> Can you try to implement a parser for debezium? We should take a look if this refactor is OK for other formats.

Currently, the pr has chosen ogg for its implementation. After the merge, I will continue to push for debezium support.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/formats/ogg/OggRecordParser.java [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+            String value = entry.getValue();
+            paimonFieldTypes.put(key, DataTypes.STRING());
+            resultMap.put(key, value);
+        }
+
+        // generate values for computed columns
+        for (ComputedColumn computedColumn : computedColumns) {
+            resultMap.put(
+                    computedColumn.columnName(),
+                    computedColumn.eval(resultMap.get(computedColumn.fieldReference())));
```
Comment (yuzelin) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/formats/ogg/OggRecordParser.java:214]: Computed column value is evaluated, should pad the paimonFieldTypes?
I also noticed that CanalRecordParser `extractRecords` has the same problem, please fix it.

**Review by yuzelin — COMMENTED** [review]
Thanks for contributing! I left some comments.

File: docs/content/how-to/cdc-ingestion.md [hunk 3]
```diff
@@ -294,11 +294,15 @@ If a message in a Kafka topic is a change event captured from another database u
...
-        <td>False</td>
+        <td>True</td>
         </tr>
     </tbody>
 </table>
 
+**Note that**
+
+In Oracle GoldenGate, the data format synchronized to Kafka does not include field data type information. As a result, Paimon sets the data type for all fields to "String" by default.
+
```
[the remark this answers is not in the document]
Comment (MonsterChenzhuo) [reply docs/content/how-to/cdc-ingestion.md:305]: done.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/KafkaSchema.java [hunk 4]
```diff
@@ -117,16 +120,13 @@ public static KafkaSchema getKafkaSchema(Configuration kafkaConfig, String topic
...
-                    if (kafkaSchema != null) {
-                        return kafkaSchema;
-                    }
-                } else {
-                    throw new UnsupportedOperationException(
-                            "This format: " + format + " is not support.");
+                DataFormat format = getDataFormat(kafkaConfig);
+                RecordParser recordParser =
+                        format.createParser(
+                                true, new TableNameConverter(true), Collections.emptyList());
```
[the remark this answers is not in the document]
Comment (MonsterChenzhuo) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/KafkaSchema.java:126]: done.

File: paimon-flink/paimon-flink-common/src/test/java/org/apache/paimon/flink/action/cdc/kafka/KafkaActionITCaseBase.java [hunk 5]
```diff
@@ -369,4 +372,19 @@ protected FileStoreTable getFileStoreTable(String tableName) throws Exception {
...
+                    try {
+                        List<String> existed = catalog().listTables(database);
+                        return existed.containsAll(Arrays.asList(tables));
+                    } catch (Catalog.DatabaseNotExistException e) {
+                        throw new RuntimeException(e);
+                    }
+                },
+                Duration.ofSeconds(50),
+                Duration.ofMillis(100),
+                "Failed to wait tables to be created in 5 seconds.");
```
[the remark this answers is not in the document]
Comment (MonsterChenzhuo) [reply paimon-flink/paimon-flink-common/src/test/java/org/apache/paimon/flink/action/cdc/kafka/KafkaActionITCaseBase.java:388]: done.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/formats/ogg/OggRecordParser.java (hunk 2 above)
Comment (MonsterChenzhuo) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/formats/ogg/OggRecordParser.java:214]: done.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/formats/ogg/OggRecordParser.java [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+    }
+
+    @Override
+    public List<RichCdcMultiplexRecord> extractRecords() {
+        List<RichCdcMultiplexRecord> records = new ArrayList<>();
+        String type = extractString(FIELD_TYPE);
+        switch (type) {
+            case OP_UPDATE:
+                processRecord(root.get(FIELD_BEFORE), RowKind.UPDATE_BEFORE, records);
+                processRecord(root.get(FIELD_AFTER), RowKind.UPDATE_AFTER, records);
```
[the remark this answers is not in the document]
Comment (MonsterChenzhuo) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/formats/ogg/OggRecordParser.java:76]: done.

**Comment by MonsterChenzhuo** [comment]
@JingsongLi @yuzelin  PTAL,thanks.

**Comment by yuzelin** [comment]
LGTM. cc @JingsongLi

**Comment by MonsterChenzhuo** [comment]
> LGTM. cc @JingsongLi
@yuzelin 
The problem of ci is a bit of a headache。I want to migrate the content of CDC into the paimon-flink/paimon-flink-cdc-e2e module in order to address the timeout issue of the utitcase-flink.yml CI process. Do you have any suggestions?

**Comment by yuzelin** [comment]
How about use syntax `mvn -T 1C -B clean install -pl 'org.apache.paimon:paimon-flink-common' -Dtest='!org.apache.paimon.flink.action.*'` in flink tests? I think this can skip the action tests. Then we can use `mvn test -Dxxx` to specify the action tests. WDYT?

**Review by JingsongLi — APPROVED** [review]
+1

**Pull request merged**